run.go

```diff
@@ -141,40 +141,80 @@ func (r *JobRunner) Init() error {
 	return nil
 }
 
+// GetDockerCreds will obtain a list of Docker credentials for the current job. This function assumes that there will be
+// at most one set of credentials for each Docker registry. The result is a map from docker registry to credentials.
+func (r *JobRunner) getDockerCreds() (map[string]*authInfo, error) {
+	result := make(map[string]*authInfo)
+
+	// Check each step in the job for credentials.
+	for _, step := range r.job.Steps {
+		container := step.Component.Container
+
+		// Add any credentials for the tool container.
+		if container.Image.Auth != "" {
+			repo := parseRepo(container.Image.Name)
+			creds, err := parse(container.Image.Auth)
+			if err != nil {
+				return nil, err
+			}
+			result[repo] = creds
+		}
+
+		// Add any credentials for data containers.
+		for _, dataContainer := range container.VolumesFrom {
+			if dataContainer.Auth != "" {
+				repo := parseRepo(dataContainer.Name)
+				creds, err := parse(dataContainer.Auth)
+				if err != nil {
+					return nil, err
+				}
+				result[repo] = creds
+			}
+		}
+	}
+
+	return result, nil
+}
+
 // DockerLogin will run "docker login" with credentials sent with the job.
 func (r *JobRunner) DockerLogin() error {
 	var err error
 	dockerBin := r.cfg.GetString("docker.path")
-	// Login so that images can be pulled.
-	var authinfo *authInfo
-	for _, img := range r.job.ContainerImages() {
-		if img.Auth != "" {
-			authinfo, err = parse(img.Auth)
-			if err != nil {
-				return err
-			}
-			authCommand := exec.Command(
-				dockerBin,
-				"login",
-				"--username",
-				authinfo.Username,
-				"--password",
-				authinfo.Password,
-				parseRepo(img.Name),
-			)
-			f, err := pty.Start(authCommand)
-			if err != nil {
-				return err
-			}
-			go func() {
-				io.Copy(logWriter, f)
-			}()
-			err = authCommand.Wait()
-			if err != nil {
-				return err
-			}
+
+	// Get credentials for each registry that requires authentication.
+	creds, err := r.getDockerCreds()
+	if err != nil {
+		return err
+	}
+
+	// Log in to the docker registres so that images can be pulled.
+	for registry, cred := range creds {
+		authCommand := exec.Command(
```

Comment:
I know this was in the prior code, but most other calls (excepting exit and upload-related ones) use `exec.CommandContext` with the context created in the main function. That ensures that if the context is canceled, it'll also cancel the command being executed. If it's not too much trouble, might be worth changing over.

Reply:
I do notice that we also don't use `pty` at all. I'm not sure if that's because it's required for this (but not anything else) or if we could also change this over to use the `command.Stderr`/`command.Stdout` like (for example) data containers do just below.

Reply:
Sounds good. I'll play with getting these changes in. Thanks for the suggestions.

Reply:
Now that I'm looking more closely, I like using `command.Stderr` and `command.Stdout` much better. Thanks again for the suggestion.